File: dsa-java/01-variables-and-data-types/hard/07_AddBinaryStrings.java
/**
 * Problem Statement:
 * Given two binary strings `a` and `b`, return their sum as a binary string.
 * The strings may be up to 10,000 characters long, far exceeding standard 64-bit primitive integers.
 * (LeetCode 67: Add Binary).
 *
 * Asked in: Facebook / Meta, Amazon, Microsoft, Apple, Bloomberg
 *
 * Approach:
 * - Two-Pointer Binary Ripple-Carry Addition:
 *   Start from the least significant bits (rightmost characters) and iterate backwards.
 *   Accumulate the sum of digits from `a`, `b`, and the existing `carry`.
 *   Current binary bit: `sum % 2`
 *   New carry: `sum / 2`
 *   Append each bit to a `StringBuilder` and reverse at the end.
 */
class AddBinaryStrings {

    public static String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int i = a.length() - 1;
        int j = b.length() - 1;
        int carry = 0;

        while (i >= 0 || j >= 0 || carry > 0) {
            int sum = carry;

            if (i >= 0) {
                sum += a.charAt(i) - '0';
                i--;
            }

            if (j >= 0) {
                sum += b.charAt(j) - '0';
                j--;
            }

            sb.append(sum % 2);
            carry = sum / 2;
        }

        return sb.reverse().toString();
    }

    public static void main(String[] args) {
        System.out.println("=== LeetCode 67: Arbitrary-Length Binary String Addition ===");

        String[][] testCases = {
            { "11", "1" },
            { "1010", "1011" },
            { "0", "0" },
            { "11111111", "1" },
            { "10000000000000000000000000000001", "11111111111111111111111111111111" }
        };

        for (String[] tc : testCases) {
            String a = tc[0];
            String b = tc[1];
            String sum = addBinary(a, b);
            System.out.printf("  %s%n+ %s%n= %s%n%n", a, b, sum);
        }
    }
}

/*
 * Time Complexity: O(max(N, M)) - Linear traversal over the length of the longer binary string.
 * Space Complexity: O(max(N, M)) - Auxiliary StringBuilder to store the resulting sum string.
 */

File: dsa-c/19-bit-manipulation/easy/07_AddBinaryStrings.c
/**
 * Problem: Add Binary (LeetCode 67)
 * Asked in: Amazon, Facebook, Microsoft
 * 
 * Given two binary strings a and b, return their sum as a binary string.
 * 
 * Approach:
 * Simulate binary addition from the rightmost bits with a carry:
 * - Start from the end of both strings and work left.
 * - sum = digitA + digitB + carry.
 * - Appended bit = sum % 2, new carry = sum / 2.
 * - Prepend each bit to result (or append and reverse at the end).
 * 
 * Time Complexity: O(max(A.length, B.length))
 * Space Complexity: O(max(A.length, B.length)) for result.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <limits.h>
#include <math.h>
#include <stdint.h>
#include <float.h>
#include <ctype.h>

#ifndef MAX
#define MAX(a, b) (((a) > (b)) ? (a) : (b))
#endif

#ifndef MIN
#define MIN(a, b) (((a) < (b)) ? (a) : (b))
#endif

static const char* toBinaryString(int n) {
    static char buf[33];
    buf[32] = '\0';
    for (int i = 31; i >= 0; i--) {
        buf[31 - i] = ((n >> i) & 1) ? '1' : '0';
    }
    return buf;
}

const char* addBinary(const char* a, const char* b) {
        StringBuilder result = new StringBuilder();
        int i = ((int)strlen(a)) - 1;
        int j = ((int)strlen(b)) - 1;
        int carry = 0;

        while (i >= 0 || j >= 0 || carry > 0) {
            int sum = carry;

            if (i >= 0) {
                sum += a[i--] - '0';
            }
            if (j >= 0) {
                sum += b[j--] - '0';
            }

            result.append(sum % 2);
            carry = sum / 2;
        }

        return result.reverse().toString();
    }

    int main(void) {
        printf("11 + 1: " + addBinary("11", "1") + " (Expected: 100)\n");
        printf("1010 + 1011: " + addBinary("1010", "1011") + " (Expected: 10101)\n");
        return 0;
}

/*
 * Time Complexity: O(N) or O(1) optimal.
 * Space Complexity: O(1) auxiliary.
 */
